src/runtime/storageNG/reference-mode-store.ts

```diff
@@ -0,0 +1,598 @@
+   */
+  private async handleProxyMessage(message: ProxyMessage<Container>): Promise<boolean> {
+    switch (message.type) {
+      case ProxyMessageType.Operations: {
+        const operations = message.operations;
+        for (const operation of operations) {
+          const entity = this.operationElement<Entity>(operation);
+          let reference = null;
+          if (entity) {
+            await this.updateBackingStore(entity);
+            reference = {id: entity.id, storageKey: this.backingStore.storageKey};
+          }
+          const containerMessage = this.updateOp<Entity, Reference>(operation, () => reference);
+          const response = await this.containerStore.onProxyMessage({type: ProxyMessageType.Operations, operations: [containerMessage], id: 1});
+          if (response) {
+            this.enqueueSend(() => void this.sendExcept(message, message.id));
+          } else {
+            return false;
+          }
+        }
+        break;
+      }
+      case ProxyMessageType.ModelUpdate: {
+        const {version, values} = message.model;
+        const newValues = {} as Dictionary<{value: Reference, version: VersionMap}>;
+        const backingStoreReceipts: Promise<boolean>[] = [];
+        Object.keys(values).forEach(id => {
+          newValues[id] = {value: {id, storageKey: this.backingStore.storageKey}, version: values[id].version};
+          backingStoreReceipts.push(this.updateBackingStore(values[id].value));
+        });
+        await Promise.all(backingStoreReceipts);
+        const model = {version, values: newValues};
+        await this.containerStore.onProxyMessage({type: ProxyMessageType.ModelUpdate, model, id: 1});
+        this.enqueueSend(() => this.sendExcept(message, message.id));
+        break;
+      }
+      case ProxyMessageType.SyncRequest: {
+        const {pendingIds, model} = this.constructPendingIdsAndModel(this.containerStore.localModel.getData());
+        const send = () => void this.callbacks.get(message.id)({type: ProxyMessageType.ModelUpdate, model: model(), id: message.id});
+        if (pendingIds.length === 0) {
+          this.enqueueSend(send);
+        } else {
+          this.enqueueBlockingSend(pendingIds, send);
+        }
+        
+        break;
+      }
+      default:
+          throw new Error('Unexpected ProxyMessageType');
+    }
+    return true;
+  }
+
+  /**
+   * Enqueues a sending function on the send queue. If the send queue is empty then
+   * the function is immediately invoked.
+   */
+  private enqueueSend(runnable: Runnable) {
+    if (this.pendingSends.length === 0) {
+      runnable();
+    } else {
+      this.pendingSends.push({fn: runnable});
+    }
+  }
+
+  /**
+   * Enqueues a send function on the send queue, deferring execution until the
+   * provided id list is available in the backing store.
+   */
+  private enqueueBlockingSend(ids: string[], runnable: Runnable) {
+    const block = Math.random() + '';
```

Comment:
Would a counter work better here instead? There doesn't seem to be a need for this to be a random number, just unique.